### PR TITLE
Feature: SourceMapDevToolPlugin progress: wip

### LINKED ISSUE #6063: Support logging statements for sourcemaps
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
Building a large codebase, sourcemaps do not provide any console logging to show that the build is in fact proceeding forward.
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
Add console logging to show the progress of the build during chunk generation
**If this is a feature request, what is motivation or use case for changing the behavior?**
The end user can confirm that the build is actually proceeding and that it is not just hanging.
**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

### PR DESCRIPTION
**Pull request by EugeneHlushko**
**What kind of change does this PR introduce?**
Displaying progress for SourceMapDevToolPlugin

**Did you add tests for your changes?**
Not yet

**If relevant, link to documentation update:**

**Summary**
Fixes #6063 

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

**Comment by EugeneHlushko** [comment]
Hi!
After pulling recent `next` i am not able to run tests locally, what do i miss?
Need to fix the tests locally before i can work on this PR further @sokra @ooflorent 

```
  963 passing (127ms)

$ mocha test/*.test.js test/*.unittest.js --max-old-space-size=4096 --harmony --trace-deprecation
module.js:515
    throw err;
    ^

Error: Cannot find module 'webpack/lib/Chunk'
    at Function.Module._resolveFilename (module.js:513:15)
    at Function.Module._load (module.js:463:25)
    at Module.require (module.js:556:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/yhl/jsfun/webpack-fork/node_modules/extract-text-webpack-plugin/dist/index.js:17:14)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Module.require (module.js:556:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/yhl/jsfun/webpack-fork/node_modules/extract-text-webpack-plugin/dist/cjs.js:3:18)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Module.require (module.js:556:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/yhl/jsfun/webpack-fork/test/HotTestCases.test.js:8:27)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Module.require (module.js:556:17)
    at require (internal/module.js:11:18)
    at /Users/yhl/jsfun/webpack-fork/node_modules/mocha/lib/mocha.js:231:27
    at Array.forEach (<anonymous>)
    at Mocha.loadFiles (/Users/yhl/jsfun/webpack-fork/node_modules/mocha/lib/mocha.js:228:14)
    at Mocha.run (/Users/yhl/jsfun/webpack-fork/node_modules/mocha/lib/mocha.js:514:10)
    at Object.<anonymous> (/Users/yhl/jsfun/webpack-fork/node_modules/mocha/bin/_mocha:480:18)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Function.Module.runMain (module.js:653:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3

```

**Comment by ooflorent** [comment]
I think you forgot `yarn link webpack`

**Comment by EugeneHlushko** [comment]
Ok it works as expected, @sokra any comments on wording and states?

File: lib/SourceMapDevToolPlugin.js [hunk 1]
```diff
@@ -82,10 +82,15 @@ class SourceMapDevToolPlugin {
...
-			compilation.hooks.afterOptimizeChunkAssets.tap("SourceMapDevToolPlugin", chunks => {
+			compilation.hooks.afterOptimizeChunkAssets.tap({
+				name: "SourceMapDevToolPlugin",
+				context: true
+			}, (context, chunks) => {
 				const moduleToSourceNameMapping = new Map();
 				const tasks = [];
+				const reportProgress = (context && context.reportProgress) ? context.reportProgress : () => {};
 
+				reportProgress(0.0, "SourceMapDevToolPlugin generation starting");
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: Don't repeat the plugin name (SourceMapDevToolPlugin) in the message, it's prepended anyway.

File: lib/SourceMapDevToolPlugin.js [hunk 2]
```diff
@@ -155,7 +160,9 @@ class SourceMapDevToolPlugin {
 					moduleToSourceNameMapping.set(module, sourceName);
 					usedNamesSet.add(sourceName);
 				}
-				tasks.forEach(task => {
+				tasks.forEach((task, index) => {
+					let progressFloat = (1.0 / tasks.length * index).toFixed(2);
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: toFixed converts a number into a string. That's not correct here. Just pass `index / tasks.length` to reportProgress that's fine.

File: lib/SourceMapDevToolPlugin.js [hunk 3]
```diff
@@ -155,7 +160,9 @@ class SourceMapDevToolPlugin {
...
+					reportProgress(progressFloat, `Generating sourcemaps ${index + 1}/${tasks.length}`);
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: `file` (`task.file`) would be a useful info here.

**Comment by sokra** [comment]
Did somebody test this? For me it looks like the expensive part happens in the first loop, but the reporting is in the second loop.

**Comment by EugeneHlushko** [comment]
Yup i did try it with small react repo looked ok from first glance. I can add couple of heavy JS files and console.time the loops just to be sure?

**Comment by EugeneHlushko** [comment]
looks pretty good thank you @sokra for joining in

**Comment by sokra** [comment]
Thanks

**Pull request merged**